src/pyrokinetics/diagnostics/saturation_rules.py

```diff
@@ -0,0 +1,247 @@
+            .sum(dim="field")
+            / heat_tot
+        )
+
+        field_squared = (
+            np.abs(eigenfunctions.where(growth_rate_tolerance < gamma_tolerance, 0.0))
+            ** 2
+        )
+
+        # Set up Jacobian and k_perp
+        k_perp = field_squared.data * 0.0
+        jacobian = field_squared.data * 0.0
+        field_factor = field_squared.data * 0.0
+
+        # Need to load MetricTerms
+        pyro.load_metric_terms(ntheta=1024)
+        theta_metric = pyro.metric_terms.regulartheta
+
+        # Add one to cover additional theta grids (like in CGYRO)
+        nperiod = pyro.numerics.nperiod + 1
+        g_tt = pyro.metric_terms.field_aligned_covariant_metric("theta", "theta")
+
+        # Can remove when we adjust outputs to handle equal_arc = True in GS2
+        if equal_arc_theta:
+            # Parallel gradient
+            grho = np.sqrt(g_tt)
+
+            # regulartheta foes from -pi to pi
+            l_theta = cumulative_trapezoid(grho, theta_metric, initial=0.0)
+            g_tt_eq = (l_theta[-1] / (2 * np.pi)) ** 2
+
+            l_theta *= 1.0 / l_theta[-1] * 2 * np.pi
+            l_theta += -np.pi
+        else:
+            l_theta = theta_metric
+
+        # Geometric theta on grid on evenly spaced l(theta) grid
+        even_l_theta = np.linspace(-np.pi, np.pi, len(l_theta))
+        theta_geo = np.interp(even_l_theta, l_theta, theta_metric)
+
+        m = np.linspace(-(nperiod - 1), nperiod - 1, 2 * nperiod - 1)
+        ntheta = len(theta_geo) - 1
+        m = np.repeat(m, ntheta)
+        theta_geo_long = np.tile(theta_geo[:-1], 2 * nperiod - 1) + 2.0 * np.pi * m
+
+        # L(theta) grid on regular theta grid long
+        m = np.linspace(-(nperiod - 1), nperiod - 1, 2 * nperiod - 1)
+        ntheta = len(l_theta) - 1
+        m = np.repeat(m, ntheta)
+        l_theta_long = np.tile(l_theta[:-1], 2 * nperiod - 1) + 2.0 * np.pi * m
+
+        for itheta0, theta0 in enumerate(theta0s.data):
+            theta_long, k_perp_long = pyro.metric_terms.k_perp(
+                ky=1.0, theta0=theta0, nperiod=nperiod
+            )
+
+            # Interp on to equal arc grid
+            if equal_arc_theta:
+                k_perp_interp = np.interp(theta, l_theta_long, k_perp_long)
+            else:
+                k_perp_interp = np.interp(theta, theta_long, k_perp_long)
+            for iky, ky in enumerate(kys.data):
+                # Technically k_perp / ky
+                k_perp[iky, itheta0, :, :] = k_perp_interp.m * ky * bunit_over_b0
+
+        bmag = pyro.metric_terms.B_magnitude
+
+        if equal_arc_theta:
+            bmag_eq_arc = np.interp(theta_geo, theta_metric, bmag)
+            bmag_long = np.tile(bmag_eq_arc[:-1], 2 * nperiod - 1)
+            pyro_jacob = pyro.metric_terms.dpsidr * np.sqrt(g_tt_eq) / bmag_long
+            bmag_balloon = np.interp(theta, l_theta_long, bmag_long)
+        else:
+            g_tt_long = np.tile(g_tt[:-1], 2 * nperiod - 1)
```

Comment:
Doesn't this `tile` approach mean the "long" grid isn't symmetric / doesn't fully cover a certain number of wraps. This might not matter, but suppose nperiod = 1 and `bmag` has three points (at theta = -pi, 0, pi) -- here we'll end up with bmag_long with just two points corresponding to theta =-pi and theta = 0. This is clearly an extreme example but if we're ever doing flux surface integrals / averages or similar with these extended arrays might we be losing a little bit of data that is actually needed for a complete answer? Again maybe these arrays are never used this way but it makes me a little nervous  (for example if someone introduces new uses of this pattern).

Reply:
So adding 1 to nperiod means you do cover the full extent of the "original" grid but it still means the total grid you end up with is missing a point so any integration over the full grid will not be quite as expected I guess? This tiling structure appears elsewhere as well.

Reply:
The aim of the tiling is to get the geometric quantities like $J$ and $|B|$ on the grid of the output simulation. These geometric quantities get calculated on a high resolution theta grid between $-\pi \rightarrow \pi$ (due to the way `MetricTerms` has been written) and then they get interpolated onto the `GKOutput` $\theta$ grid. All calculations including any sort of flux surface average is done on that grid, not the high resolution grid used to make the geometric quantities.

Reply:
Ah OK that makes sense. I came across it looking at the k_perp metric terms method as Arka noticed the length of the output grid was not as expected. Scanning the code I think `k_perp` is actually the _only_ place using the tile structure apart from this new PR. Is there any danger about people misusing the k_perp output?

Reply:
Probably yes, its probably quite easy to fix to be honest.

To clarify what does GS2 do with its grid? If you have `ntheta` then from $-\pi \rightarrow \pi$ (inclusive of $\pi$) will have `ntheta+1` points? In total there are $\bigg((2n_{period} - 1) * n_{theta}\bigg) + 1$ points? 

That $+1$ is different to say CGYRO or GENE which just have $\bigg((2n_{period} - 1) * n_{theta}\bigg)$ points. I am happy to add on the final point but I am not sure what the best strategy is here...

Reply:
A note that GENE/CGYRO do include 0, but maintain the multiple of `ntheta` grid points by dropping the "final" theta grid point as we do here, meaning the final point is not an integer multiple of $\pi$ so $\theta_{final} mod 2\pi \ne 0$

Reply:
Yes in GS2 we have `ntheta+1` points spanning the inclusive range from -pi to pi (and each subsequent 2 pi section). This means that in ballooning space we always have `theta(one_end) == - theta(other_end)`. This means that we always cover the poloidal section exactly an integer number of times. Provided the theta grid used in integrals is consistent then I guess it is unlikely to cause a major issue, but considering low resolution, low extent cases I could see there being slight errors in flux surface integrals when you drop the last point (without introducing any special handling).

Reply:
If you want to match a code exactly so you model the same `theta` points then you may need to set up metric terms with one extra point i.e. to account for how the code sets up its theta grid. This would be the case in GS2 I think

```python
pyro.load_metric_terms(ntheta=pyro.numerics.ntheta + 1)
```